### PR TITLE
Update push docker container to support container registry

### PR DESCRIPTION
**Pull request by pfleidi**
The changes in this PR update the `docker-publish` workflow in a few ways:

* It now supports logging in and pushing images to the [GH Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry)
* Instead of using error-prone shell scripts to extract metadata and build/push images it's using Docker actions
* The registry being used is now defined in a variable

## Pre-requisites

- [x] Prior to submitting a new workflow, please apply to join the GitHub Technology Partner Program: [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner).
  * I'm a GitHub employee so I'm assuming that doesn't apply to me 

## Tasks

**For _all_ workflows, the workflow:**

- [x] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [x] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [x] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [x] Should include comments in the workflow for any parts that are not obvious or could use clarification.

**For _CI_ workflows, the workflow:**

- [x] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [x] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [x] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [ ] Packaging workflows should run on `release` with `types: [ created ]`.
- [x] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**Some general notes:**

- [ ] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [x] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [x] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [x] Automation and CI workflows cannot be dependent on a paid service or product.

### CODE REVIEW

File: ci/docker-publish.yml
Comment (clarkbw) [inline ci/docker-publish.yml:66]: This doesn't actually matter because the service doesn't check it but don't we want `github.actor` here?

```suggestion
          username: ${{ github.actor }}
```

File: ci/docker-publish.yml
Comment (clarkbw) [inline ci/docker-publish.yml:82]: If we can drop this and it defaults to `.` let's do that.

File: ci/docker-publish.yml
Comment (clarkbw) [inline ci/docker-publish.yml:84]: Unless this is the Dependabot issue do we need to prevent pushing from PRs?

File: ci/docker-publish.yml
Comment (clarkbw) [inline ci/docker-publish.yml:77]: It'd be best if we could get this down to zero elements. The `semver` seems required and hopefully we can change upstream to fix that.

File: ci/docker-publish.yml
Comment (pfleidi) [reply ci/docker-publish.yml:82]: I'll see if we can skip it and drop it if it does.

File: ci/docker-publish.yml
Comment (pfleidi) [reply ci/docker-publish.yml:84]: I was assuming that most developers would not want that because it would slow down their CI feedback circle. Depending on your CI pipeline, it could be pretty wasteful in terms of time and traffic to push every CI build to a registry so I assumed it was a sensible default.

In which scenarios would we want PRs to be pushed as well?

File: ci/docker-publish.yml
Comment (pfleidi) [reply ci/docker-publish.yml:77]: @clarkbw Do we want to track an issue somewhere to add the changes to upstream?

Comment (pfleidi) [reply ci/docker-publish.yml:77]: I've tried to slim it down to only the semver version but pushing to anything but a tag failed because the push operation needs a tag.

It failed with this message:

`Error: buildx call failed with: error: tag is needed when pushing to registry`

I feel like we need to take this upstream if we wanted to minimize the tags configuration.

Comment (pfleidi) [reply ci/docker-publish.yml:77]: Also: When removing the `event=pr` type, it technically works but we're getting warnings as well:

image

File: ci/docker-publish.yml
Comment (clarkbw) [reply ci/docker-publish.yml:83]: I don't think the push has a high cost in terms of time, if we've already built the container as part of our PR we might as well push it and any layers will be reused in a future build after merging.

File: ci/docker-publish.yml
Comment (crazy-max) [reply ci/docker-publish.yml:77]: @pfleidi It's safe to remove the `tags` input for the starter workflow. [Default entries if empty](https://github.com/docker/metadata-action#tags-input) should be enough.

File: ci/docker-publish.yml
Comment (crazy-max) [inline ci/docker-publish.yml:77]: @pfleidi It's safe to remove the `tags` input for the starter workflow. [Default entries if empty](https://github.com/docker/metadata-action#tags-input) should be enough.

File: ci/docker-publish.yml
Comment (crazy-max) [reply ci/docker-publish.yml:83]: For GHCR it would be fine to push an image but would need to be done through `pull_request_target` event to be able to use the `GITHUB_TOKEN` secret right?

File: ci/docker-publish.yml
Comment (crazy-max) [reply ci/docker-publish.yml]: @clarkbw About that, it's true when using a GITHUB_TOKEN but not for a PAT right? I think GITHUB_TOKEN usage is the right move so do we need to also [update the example on docker-login action repo](https://github.com/docker/login-action#github-container-registry)?

File: ci/docker-publish.yml
Comment (clarkbw) [reply ci/docker-publish.yml:83]: Yes, though the odd thing about the `pull_request_target` is that it pulls the target branch so you're not actually building the PR branch. This makes the artifacts created somewhat useless.  Ex: If your PR branch `foo` is targeting `main` `pull_request_target` will do a build against `main` instead of `foo`; which isn't what we want. However if we use the `pull_request` and the user who created the PR doesn't have write access the push will fail because the `GITHUB_TOKEN` is read-only.

Comment (crazy-max) [reply ci/docker-publish.yml:83]: Ok so the condition on pull_request to push is correct here.

**Comment by dominic59595** [comment]
pfleidi:packages-container-registry-workflow

**Comment by dominic59595** [comment]
[pfleidi:packages-container-registry-workflow]() [](pfleidi:packages-container-registry-workflow) # #
Duplicate of #

**Comment by dominic59595** [comment]
+
 on:
   push:
     branches: [ $default-branch ]

From 41a66c656e20c398aa7143b649d3678dae503afc Mon Sep 17 00:00:00 2001
From: Sven Pfleiderer <pfleidi@github.com>
Date: Wed, 26 May 2021 13:46:54 -0700
Subject: [PATCH 3/5] Use git shas rather than version tags

---
 ci/docker-publish.yml | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/ci/docker-publish.yml b/ci/docker-publish.yml
index 63b02f26e..26a5f89fa 100644
--- a/ci/docker-publish.yml
+++ b/ci/docker-publish.yml
@@ -60,7 +60,7 @@ jobs:
         uses: actions/checkout@v2
 
       - name: Log into registry ${{ env.REGISTRY }}
-        uses: docker/login-action@v1
+        uses: docker/login-action@28218f9b04b4f3f62068d7b6ce6ca5b26e35336c
         with:
           registry: ${{ env.REGISTRY }}
           username: ${{ github.repository_owner }}
@@ -68,7 +68,7 @@ jobs:
 
       - name: Extract Docker metadata
         id: meta
-        uses: docker/metadata-action@v3
+        uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
         with:
           images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
           tags: |
@@ -77,7 +77,7 @@ jobs:
               type=semver,pattern={{version}}
 
       - name: Build and push Docker image
-        uses: docker/build-push-action@v2
+        uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
         with:
           context: .
           # Don't push on pull requests

From aa30f1448e6f181d695a784d284dbeaefc12a208 Mon Sep 17 00:00:00 2001
From: Sven Pfleiderer <sg+github@roothausen.de>
Date: Wed, 26 May 2021 14:07:29 -0700
Subject: [PATCH 4/5] Update ci/docker-publish.yml

Co-authored-by: Bryan Clark <clarkbw@github.com>
---
 ci/docker-publish.yml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ci/docker-publish.yml b/ci/docker-publish.yml
index 26a5f89fa..d2672ab07 100644
--- a/ci/docker-publish.yml
+++ b/ci/docker-publish.yml
@@ -63,7 +63,7 @@ jobs:
         uses: docker/login-action@28218f9b04b4f3f62068d7b6ce6ca5b26e35336c
         with:
           registry: ${{ env.REGISTRY }}
-          username: ${{ github.repository_owner }}
+          username: ${{ github.actor }}
           password: ${{ secrets.GITHUB_TOKEN }}
 
       - name: Extract Docker metadata

From 6fe53a9ebd963520a1da2b470d71380c14c284fb Mon Sep 17 00:00:00 2001
From: Sven Pfleiderer <pfleidi@github.com>
Date: Wed, 26 May 2021 15:03:41 -0700
Subject: [PATCH 5/5] Remove context configuration since it's already the
 default

---
 ci/docker-publish.yml | 1 -
 1 file changed, 1 deletion(-)

diff --git a/ci/docker-publish.yml b/ci/docker-publish.yml
index 26a5f89fa..c820f9177 100644
--- a/ci/docker-publish.yml
+++ b/ci/docker-publish.yml
@@ -79,7 +79,6 @@ jobs:
       - name: Build and push Docker image
         uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
         with:
-          context: .
           # Don't push on pull requests
           push: ${{ github.event_name != 'pull_request' }}
           tags: ${{ steps.meta.outputs.tags }}

File: ci/docker-publish.yml
Comment (pfleidi) [reply ci/docker-publish.yml]: I've removed that condition for now.

**Review by andymckay — APPROVED** [review]

**Pull request merged**
